### PR TITLE
llama : sanity checks for access to logits

### PR DESCRIPTION
**Pull request by cebtenzzre**
This change attempts to validate logit access in debug mode. No effect on release builds.

This change is incomplete:
- We can't actually validate access via llama_get_logits (which is the problem I actually had when transitioning to the new batch API - I wasn't using llama_get_logits_ith), because it has two meanings. Most people use the undocumented meaning: 1D array of logits for the final token.
- Since we can't identify which meaning the programmer intends when they call llama_get_logits, could we deprecate it and provide a replacement that has never and will never return a 1D array, e.g. "llama_get_logits_all"? This would avoid a footgun when developers transition to the new batch API, making it clear to them via compiler warnings (we should actually use `__attribute__((deprecated))`) that they need to use llama_get_logits_ith.
- I'm assuming we don't intend to keep llama_eval and llama_batch_get_one forever, so those should be properly deprecated as well so people don't use them in new code.

cc @ggerganov

### CODE REVIEW

File: llama.cpp [hunk 1]
```diff
@@ -9465,6 +9484,7 @@ float * llama_get_logits(struct llama_context * ctx) {
 }
 
 float * llama_get_logits_ith(struct llama_context * ctx, int32_t i) {
+    assert(ctx->logits_valid.at(i));
```
Comment (ggerganov) [inline llama.cpp:9487]: Should this be:

```suggestion
    assert(ctx->logits_valid.at(i*ctx->model.hparams.n_vocab));
```

?

File: llama.cpp [hunk 2]
```diff
@@ -5609,20 +5613,35 @@ static int llama_decode_internal(
     {
         auto & logits_out = lctx.logits;
 
+#ifndef NDEBUG
+        auto & logits_valid = lctx.logits_valid;
+        logits_valid.clear();
+        logits_valid.resize(n_vocab);
```
Comment (ggerganov) [inline llama.cpp:5619]: Should this be this then?

```suggestion
        logits_valid.resize(n_tokens);
```

Comment (cebtenzzre) [reply llama.cpp]: Yep, thanks.

**Comment by cebtenzzre** [comment]
> In which cases `llama_get_logits_all` would be useful?

I assumed that since it is something you can do with the API then somebody had a use for it 😛 

I guess it makes sense that llama_get_logits_ith can always be used, and with this PR it's a safer way to do it.

Should I deprecate the necessary functions in this PR or a future one?

**Comment by cmp-nct** [comment]
I am using llama_get_logits() also to write logits back from a vector into the original data.
There is currently no other API function that would support that, if the function gets abstracted or changed it would be good to add a llama_write_logits() so no functionality is lost

**Comment by cebtenzzre** [comment]
> Just as a sidenote: I am using llama_get_logits_ith() also to write logits back from a vector into the original data.

What are the dimensions of logits_backup? My best guess is that you want to be able to write logits for one token that has not had any logits computed. If a function to write to the logits for a token at index i wouldn't be less convenient, I think we could add that. Nobody expects the previous logits to persist after a further call to llama_decode, right?

**Review by ggerganov — APPROVED** [review]
> > In which cases `llama_get_logits_all` would be useful?
> 
> I assumed that since it is something you can do with the API then somebody had a use for it 😛
> 
> I guess it makes sense that llama_get_logits_ith can always be used, and with this PR it's a safer way to do it.
> 
> Should I deprecate the necessary functions in this PR or a future one?

Let's merge this PR and deprecate in a new one.

Regarding `llama_get_logits_all` - yup, let's add it.

> writing logits

This still remains an option, though not officially recommended by the API. I.e. could break in the future, but likely it won't.
If we decide to break this in the future - then we will add `llama_set_logits`. But for the moment, I don't think we should

**Comment by cebtenzzre** [comment]
> This still remains an option, though not officially recommended by the API. I.e. could break in the future, but likely it won't.

This PR does break it though, because it prevents you from reading _or_ writing logits that were not recently written to during eval. cmp-nct wants to be able to sample without any previous eval. We have two options for supporting it:
- Remove the assertions and just rely on zeroing the vector so access to incorrect logits is least deterministic - in this case we should also zero the vector in release builds for consistent behavior, since there would be no obvious way to detect abuse.
- Add an API to set the logits, which will mark them as valid.

**Comment by bullno1** [comment]
Should `llama_context_params.logits_all` be deprecated too?

CMIIW but the preferred way is to set all `llama_batch.logits` to 1.

**Comment by cebtenzzre** [comment]
This seems like a step in the right direction, so I'll merge it.

File: llama.cpp [hunk 3]
```diff
@@ -5609,20 +5613,37 @@ static int llama_decode_internal(
...
             logits_out.resize(n_vocab * n_tokens);
             memcpy(logits_out.data(), (float *) ggml_get_data(res), sizeof(float)*n_vocab*n_tokens);
+#ifndef NDEBUG
+            std::fill(logits_valid.begin(), logits_valid.end(), true);
+#endif
         } else {
             logits_out.resize(n_vocab);
             memcpy(logits_out.data(), (float *) ggml_get_data(res) + (n_vocab*(n_tokens - 1)), sizeof(float)*n_vocab);
+#ifndef NDEBUG
+            logits_valid[n_tokens - 1] = true;
```
Comment (slaren) [inline llama.cpp:5645]: Shouldn't this be `logits_valid[0] = true;`? Currently `main` crashes on debug builds when using any prompt.

Comment (cebtenzzre) [reply llama.cpp:5645]: #4516
